File: Bcore/src/main/java/top/niunaijun/blackbox/core/web/ShadowWebViewActivity.java
package top.niunaijun.blackbox.core.web;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.Log;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.ArrayList;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import top.niunaijun.blackbox.R;

public class ShadowWebViewActivity extends AppCompatActivity {
    public static final String EXTRA_ORIGINAL_INTENT = "_B_|_shadow_web_original_intent_";
    public static final String EXTRA_TARGET_PACKAGE = "_B_|_shadow_web_target_pkg_";
    public static final String EXTRA_TARGET_CLASS = "_B_|_shadow_web_target_cls_";
    private static final String TAG = "ShadowWebViewActivity";

    private WebView webView;
    private ProgressBar progressBar;
    private TextView statusView;
    private Button openExternalButton;
    private Button closeButton;
    private String targetUrl;

    public static Intent createIntent(String hostPackage, @Nullable Intent original) {
        Intent intent = new Intent();
        intent.setClassName(hostPackage, ShadowWebViewActivity.class.getName());
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.putExtra(EXTRA_ORIGINAL_INTENT, original);
        if (original != null && original.getComponent() != null) {
            intent.putExtra(EXTRA_TARGET_PACKAGE, original.getComponent().getPackageName());
            intent.putExtra(EXTRA_TARGET_CLASS, original.getComponent().getClassName());
        }
        return intent;
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_shadow_webview_core);

        webView = findViewById(R.id.shadow_webview);
        progressBar = findViewById(R.id.shadow_web_progress);
        statusView = findViewById(R.id.shadow_web_status);
        openExternalButton = findViewById(R.id.shadow_web_open_external);
        closeButton = findViewById(R.id.shadow_web_close);

        final Intent originalIntent = getParcelableIntentExtra(getIntent(), EXTRA_ORIGINAL_INTENT);
        targetUrl = UrlExtractor.extract(originalIntent);

        if (getSupportActionBar() != null) {
            String title = originalIntent != null && originalIntent.getComponent() != null
                    ? originalIntent.getComponent().getShortClassName() : "Web";
            getSupportActionBar().setTitle(title);
        }

        openExternalButton.setOnClickListener(v -> {
            if (targetUrl != null) {
                openExternal(targetUrl);
            }
        });
        closeButton.setOnClickListener(v -> finish());

        if (targetUrl == null || targetUrl.trim().isEmpty()) {
            statusView.setText("未从目标 Intent 中解析到网页地址。\n\n目标类："
                    + getIntent().getStringExtra(EXTRA_TARGET_CLASS)
                    + "\n\nExtras:\n"
                    + UrlExtractor.dumpExtras(originalIntent));
            progressBar.setVisibility(View.GONE);
            webView.setVisibility(View.GONE);
            openExternalButton.setVisibility(View.GONE);
            return;
        }

        statusView.setText(targetUrl);
        CookieManager.getInstance().setAcceptCookie(true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            CookieManager.getInstance().setAcceptThirdPartyCookies(webView, true);
            webView.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        }

        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setAllowFileAccess(true);
        settings.setAllowContentAccess(true);
        settings.setSupportZoom(true);
        settings.setBuiltInZoomControls(false);
        settings.setDisplayZoomControls(false);
        settings.setLoadsImagesAutomatically(true);
        settings.setMediaPlaybackRequiresUserGesture(false);
        settings.setCacheMode(WebSettings.LOAD_DEFAULT);
        settings.setJavaScriptCanOpenWindowsAutomatically(true);
        settings.setUserAgentString(settings.getUserAgentString() + " BlackBoxShadowWeb/2.0");

        GenericJavascriptBridge bridge = new GenericJavascriptBridge();
        webView.addJavascriptInterface(bridge, "javascript");
        webView.addJavascriptInterface(bridge, "KRJavascriptInterface");
        webView.addJavascriptInterface(bridge, "Android");

        webView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView view, int newProgress) {
                progressBar.setProgress(newProgress);
                progressBar.setVisibility(newProgress >= 100 ? View.GONE : View.VISIBLE);
            }
        });
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                statusView.setText(url != null ? url : targetUrl);
                progressBar.setVisibility(View.VISIBLE);
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                String url = request != null && request.getUrl() != null ? request.getUrl().toString() : null;
                return handleSpecialUrl(url);
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                return handleSpecialUrl(url);
            }
        });

        Log.d(TAG, "Loading url: " + targetUrl + " from " + (originalIntent != null ? originalIntent.getComponent() : null));
        webView.loadUrl(targetUrl);
    }

    @SuppressWarnings("deprecation")
    @Nullable
    private static Intent getParcelableIntentExtra(Intent intent, String key) {
        if (intent == null) return null;
        try {
            return (Intent) intent.getParcelableExtra(key);
        } catch (Throwable ignored) {
            return null;
        }
    }

    private boolean handleSpecialUrl(String url) {
        if (url == null || url.trim().isEmpty()) return false;
        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            openExternal(url);
            return true;
        }
        return false;
    }

    private void openExternal(String url) {
        try {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
        } catch (ActivityNotFoundException e) {
            statusView.setText("无法打开外部浏览器：" + url + "\n" + e.getMessage());
        }
    }

    @Override
    public void onBackPressed() {
        if (webView != null && webView.getVisibility() == View.VISIBLE && webView.canGoBack()) {
            webView.goBack();
            return;
        }
        super.onBackPressed();
    }

    @Override
    protected void onDestroy() {
        if (webView != null) {
            webView.stopLoading();
            webView.removeAllViews();
            webView.destroy();
        }
        super.onDestroy();
    }

    public class GenericJavascriptBridge {
        @JavascriptInterface public void close() { runOnUiThread(() -> finish()); }
        @JavascriptInterface public void back() { runOnUiThread(() -> onBackPressed()); }
        @JavascriptInterface public void open(String url) {
            if (url != null && !url.trim().isEmpty()) {
                runOnUiThread(() -> {
                    if (!handleSpecialUrl(url)) {
                        webView.loadUrl(url);
                    }
                });
            }
        }
        @JavascriptInterface public void postMessage(String message) { Log.d(TAG, "postMessage=" + message); }
        @JavascriptInterface public void callback(String message) { Log.d(TAG, "callback=" + message); }
        @JavascriptInterface public String getDeviceInfo() { return "{}"; }
        @JavascriptInterface public String getOaid() { return ""; }
        @JavascriptInterface public String getOAID() { return ""; }
        @JavascriptInterface public String getUserInfo() { return "{}"; }
    }

    private static class UrlExtractor {
        private static final Pattern URL_PATTERN = Pattern.compile("https?://[^\\s\"'<>]+", Pattern.CASE_INSENSITIVE);
        private static final String[] COMMON_KEYS = new String[] {
                "url", "URL", "link", "web_url", "webUrl", "openUrl", "pageUrl", "h5Url", "jumpUrl", "targetUrl", "loadUrl"
        };

        @Nullable
        static String extract(@Nullable Intent intent) {
            if (intent == null) return null;
            String dataString = intent.getDataString();
            String fromString = extractFromString(dataString);
            if (fromString != null) return fromString;
            for (String key : COMMON_KEYS) {
                String extra = intent.getStringExtra(key);
                String url = extractFromString(extra);
                if (url != null) return url;
            }
            return extractFromBundle(intent.getExtras());
        }

        static String dumpExtras(@Nullable Intent intent) {
            return dumpBundle(intent != null ? intent.getExtras() : null);
        }

        @Nullable
        private static String extractFromBundle(@Nullable Bundle bundle) {
            if (bundle == null) return null;
            Set<String> keys = bundle.keySet();
            for (String key : keys) {
                Object value = bundle.get(key);
                String url = extractFromAny(value);
                if (url != null) return url;
            }
            return null;
        }

        @Nullable
        private static String extractFromAny(@Nullable Object value) {
            if (value == null) return null;
            if (value instanceof String) return extractFromString((String) value);
            if (value instanceof CharSequence) return extractFromString(value.toString());
            if (value instanceof Uri) return extractFromString(value.toString());
            if (value instanceof Intent) return extract((Intent) value);
            if (value instanceof Bundle) return extractFromBundle((Bundle) value);
            if (value instanceof ArrayList) {
                for (Object item : (ArrayList<?>) value) {
                    String url = extractFromAny(item);
                    if (url != null) return url;
                }
                return null;
            }
            if (value instanceof Object[]) {
                for (Object item : (Object[]) value) {
                    String url = extractFromAny(item);
                    if (url != null) return url;
                }
                return null;
            }
            return extractFromString(String.valueOf(value));
        }

        @Nullable
        private static String extractFromString(@Nullable String raw) {
            if (raw == null) return null;
            String trimmed = raw.trim();
            if (trimmed.isEmpty()) return null;
            Matcher matcher = URL_PATTERN.matcher(trimmed);
            return matcher.find() ? matcher.group() : null;
        }

        static String dumpBundle(@Nullable Bundle bundle) {
            if (bundle == null) return "<empty>";
            StringBuilder sb = new StringBuilder();
            for (String key : bundle.keySet()) {
                Object value = bundle.get(key);
                sb.append(String.format(Locale.US, "%s = %s", key, value)).append('\n');
            }
            return sb.length() == 0 ? "<empty>" : sb.toString();
        }
    }
}

File: app/src/main/java/top/niunaijun/blackboxa/view/web/ShadowWebViewActivity.kt
package top.niunaijun.blackboxa.view.web

import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.webkit.CookieManager
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import top.niunaijun.blackboxa.R
import java.util.Locale
import java.util.regex.Pattern

class ShadowWebViewActivity : AppCompatActivity() {

    companion object {
        const val EXTRA_ORIGINAL_INTENT = "_B_|_shadow_web_original_intent_"
        const val EXTRA_TARGET_PACKAGE = "_B_|_shadow_web_target_pkg_"
        const val EXTRA_TARGET_CLASS = "_B_|_shadow_web_target_cls_"
        private const val TAG = "ShadowWebViewActivity"

        @JvmStatic
        fun createIntent(hostPackage: String, original: Intent?): Intent {
            return Intent().apply {
                setClassName(hostPackage, ShadowWebViewActivity::class.java.name)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                putExtra(EXTRA_ORIGINAL_INTENT, original)
                putExtra(EXTRA_TARGET_PACKAGE, original?.component?.packageName)
                putExtra(EXTRA_TARGET_CLASS, original?.component?.className)
            }
        }
    }

    private lateinit var webView: WebView
    private lateinit var progressBar: ProgressBar
    private lateinit var statusView: TextView
    private lateinit var openExternalButton: Button
    private lateinit var closeButton: Button

    private var targetUrl: String? = null

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_shadow_webview)

        webView = findViewById(R.id.shadow_webview)
        progressBar = findViewById(R.id.shadow_web_progress)
        statusView = findViewById(R.id.shadow_web_status)
        openExternalButton = findViewById(R.id.shadow_web_open_external)
        closeButton = findViewById(R.id.shadow_web_close)

        val originalIntent: Intent? = intent.getParcelableExtra(EXTRA_ORIGINAL_INTENT)
        targetUrl = UrlExtractor.extract(originalIntent)

        supportActionBar?.title = originalIntent?.component?.shortClassName ?: "Web"

        openExternalButton.setOnClickListener {
            targetUrl?.let { url -> openExternal(url) }
        }
        closeButton.setOnClickListener { finish() }

        if (targetUrl.isNullOrBlank()) {
            statusView.text = buildString {
                append("未从目标 Intent 中解析到网页地址。\n\n")
                append("目标类：")
                append(intent.getStringExtra(EXTRA_TARGET_CLASS) ?: "unknown")
                append("\n\nExtras:\n")
                append(UrlExtractor.dumpExtras(originalIntent))
            }
            progressBar.visibility = View.GONE
            webView.visibility = View.GONE
            openExternalButton.visibility = View.GONE
            return
        }

        statusView.text = targetUrl
        CookieManager.getInstance().setAcceptCookie(true)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            CookieManager.getInstance().setAcceptThirdPartyCookies(webView, true)
            webView.settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }

        with(webView.settings) {
            javaScriptEnabled = true
            domStorageEnabled = true
            databaseEnabled = true
            allowFileAccess = true
            allowContentAccess = true
            setSupportZoom(true)
            builtInZoomControls = false
            displayZoomControls = false
            loadsImagesAutomatically = true
            mediaPlaybackRequiresUserGesture = false
            cacheMode = WebSettings.LOAD_DEFAULT
            javaScriptCanOpenWindowsAutomatically = true
            userAgentString = userAgentString + " BlackBoxShadowWeb/1.0"
        }

        webView.addJavascriptInterface(GenericJavascriptBridge(), "javascript")
        webView.addJavascriptInterface(GenericJavascriptBridge(), "KRJavascriptInterface")
        webView.addJavascriptInterface(GenericJavascriptBridge(), "Android")

        webView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                progressBar.progress = newProgress
                progressBar.visibility = if (newProgress >= 100) View.GONE else View.VISIBLE
            }
        }
        webView.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                statusView.text = url ?: targetUrl
                progressBar.visibility = View.VISIBLE
            }

            override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                val url = request?.url?.toString() ?: return false
                return handleSpecialUrl(url)
            }

            override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                return handleSpecialUrl(url)
            }
        }

        Log.d(TAG, "Loading url: $targetUrl from ${originalIntent?.component}")
        webView.loadUrl(targetUrl!!)
    }

    private fun handleSpecialUrl(url: String?): Boolean {
        if (url.isNullOrBlank()) return false
        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            openExternal(url)
            return true
        }
        return false
    }

    private fun openExternal(url: String) {
        try {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
        } catch (e: ActivityNotFoundException) {
            statusView.text = "无法打开外部浏览器：$url\n${e.message}"
        }
    }

    override fun onBackPressed() {
        if (::webView.isInitialized && webView.visibility == View.VISIBLE && webView.canGoBack()) {
            webView.goBack()
            return
        }
        super.onBackPressed()
    }

    override fun onDestroy() {
        if (::webView.isInitialized) {
            webView.stopLoading()
            webView.removeAllViews()
            webView.destroy()
        }
        super.onDestroy()
    }

    inner class GenericJavascriptBridge {
        @JavascriptInterface fun close() = runOnUiThread { finish() }
        @JavascriptInterface fun back() = runOnUiThread { onBackPressed() }
        @JavascriptInterface fun open(url: String?) {
            if (!url.isNullOrBlank()) {
                runOnUiThread {
                    if (!handleSpecialUrl(url)) {
                        webView.loadUrl(url)
                    }
                }
            }
        }
        @JavascriptInterface fun postMessage(message: String?) { Log.d(TAG, "postMessage=$message") }
        @JavascriptInterface fun callback(message: String?) { Log.d(TAG, "callback=$message") }
        @JavascriptInterface fun getDeviceInfo(): String = "{}"
        @JavascriptInterface fun getOaid(): String = ""
        @JavascriptInterface fun getOAID(): String = ""
        @JavascriptInterface fun getUserInfo(): String = "{}"
    }

    private object UrlExtractor {
        private val URL_PATTERN = Pattern.compile("https?://[^\\s\"'<>]+", Pattern.CASE_INSENSITIVE)
        private val COMMON_KEYS = listOf(
            "url", "URL", "link", "web_url", "webUrl", "openUrl", "pageUrl", "h5Url", "jumpUrl", "targetUrl", "loadUrl"
        )

        fun extract(intent: Intent?): String? {
            if (intent == null) return null
            intent.dataString?.let { extractFromString(it)?.let { url -> return url } }
            for (key in COMMON_KEYS) {
                intent.getStringExtra(key)?.let { extractFromString(it)?.let { url -> return url } }
            }
            return extractFromBundle(intent.extras)
        }

        fun dumpExtras(intent: Intent?): String {
            return dumpBundle(intent?.extras)
        }

        private fun extractFromBundle(bundle: Bundle?): String? {
            if (bundle == null) return null
            for (key in bundle.keySet()) {
                when (val value = bundle.get(key)) {
                    is String -> extractFromString(value)?.let { return it }
                    is CharSequence -> extractFromString(value.toString())?.let { return it }
                    is Uri -> extractFromString(value.toString())?.let { return it }
                    is Intent -> extract(value)?.let { return it }
                    is Bundle -> extractFromBundle(value)?.let { return it }
                    is ArrayList<*> -> value.forEach { item -> extractFromAny(item)?.let { return it } }
                    is Array<*> -> value.forEach { item -> extractFromAny(item)?.let { return it } }
                    is Parcelable -> extractFromString(value.toString())?.let { return it }
                    else -> extractFromString(value?.toString())?.let { return it }
                }
            }
            return null
        }

        private fun extractFromAny(value: Any?): String? = when (value) {
            null -> null
            is String -> extractFromString(value)
            is CharSequence -> extractFromString(value.toString())
            is Uri -> extractFromString(value.toString())
            is Intent -> extract(value)
            is Bundle -> extractFromBundle(value)
            else -> extractFromString(value.toString())
        }

        private fun extractFromString(raw: String?): String? {
            if (raw.isNullOrBlank()) return null
            val trimmed = raw.trim()
            val matcher = URL_PATTERN.matcher(trimmed)
            return if (matcher.find()) matcher.group() else null
        }

        private fun dumpBundle(bundle: Bundle?): String {
            if (bundle == null) return "<empty>"
            val lines = mutableListOf<String>()
            for (key in bundle.keySet()) {
                val value = bundle.get(key)
                lines += String.format(Locale.US, "%s = %s", key, value)
            }
            return if (lines.isEmpty()) "<empty>" else lines.joinToString("\n")
        }
    }
}
